diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,320 @@
+        outer = jtu.tree_structure(gamma)
+        inner = self._coeffs_structure
+        coeffs_with_tree = jtu.tree_transpose(outer, inner, tree_with_coeffs)
+        return coeffs_with_tree
+
+    def init(
+        self,
+        terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: LangevinTuple,
+        args: PyTree,
+    ) -> SolverState:
+        """Precompute _SolverState which carries the Taylor coefficients and the
+        SRK coefficients (which can be computed from h and the Taylor coefficients).
+        Some solvers of this type are FSAL, so _SolverState also carries the previous
+        evaluation of grad_f.
+        """
+        drift, diffusion = terms.terms
+        gamma, u, f = get_args_from_terms(terms)
+
+        h = drift.contr(t0, t1)
+        x0, v0 = y0
+
+        def _check_shapes(_c, _u, _x, _v):
+            assert _c.shape == _u.shape == _x.shape == _v.shape
+
+        assert jtu.tree_all(jtu.tree_map(_check_shapes, gamma, u, x0, v0))
+
+        tay_coeffs = jtu.tree_map(self._comp_taylor_coeffs_leaf, gamma)
+        # tay_coeffs have the same tree structure as gamma, with each leaf being a
+        # _Coeffs and the arrays have an extra trailing dimension of 6
+
+        coeffs = self._recompute_coeffs(h, gamma, tay_coeffs, -1.0)
+        rho = jtu.tree_map(lambda c, _u: jnp.sqrt(2 * c * _u), gamma, u)
+
+        state_out = SolverState(
+            h=h,
+            taylor_coeffs=tay_coeffs,
+            coeffs=coeffs,
+            rho=rho,
+            prev_f=f(x0),
+        )
+
+        return state_out
+
+    @staticmethod
+    @abc.abstractmethod
+    def _compute_step(
+        h: RealScalarLike,
+        levy,
+        x0: LangevinX,
+        v0: LangevinX,
+        langevin_args: _LangevinArgs,
+        coeffs: _Coeffs,
+        st: SolverState,
+    ) -> tuple[LangevinX, LangevinX, LangevinX, _ErrorEstimate]:
+        raise NotImplementedError
+
+    def step(
+        self,
+        terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: LangevinTuple,
+        args: PyTree,
+        solver_state: SolverState,
+        made_jump: BoolScalarLike,
+    ) -> tuple[LangevinTuple, _ErrorEstimate, DenseInfo, SolverState, RESULTS]:
+        del made_jump, args
+        st = solver_state
+        drift, diffusion = terms.terms
+        gamma, u, f = get_args_from_terms(terms)
+
+        h = drift.contr(t0, t1)
+        h_prev = st.h
+        tay: PyTree[_Coeffs] = st.taylor_coeffs
+        coeffs: _Coeffs = st.coeffs
+
+        # If h changed recompute coefficients
+        # Even when using constant step sizes, h can fluctuate by small amounts,
+        # so we use `jnp.isclose` for comparison
+        cond = jnp.isclose(h_prev, h, rtol=1e-10, atol=1e-12)
+        coeffs = lax.cond(
+            cond,
```

Comment:
Optimization: make this `eqxi.unvmap_all(cond)`.

Reply:
Makes sense.